File: src/main/java/graveler/action/UpdateStress.java
package graveler.action;

import graveler.util.PointedWorld;
import graveler.util.WorldExtensions;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;

public record UpdateStress(BlockPos pos) implements Action {
  @Override
  public double weight() {
    return 0.0;
  }

  @Override
  public void apply(ActionContext context) {
    PointedWorld point = WorldExtensions.pointedAt(context.world(), pos);

    if (!point.isStressAware()) {
      point.clearStress();
      return;
    }

    if (!context.world().hasChunkAt(pos)) {
      return;
    }

    int newStress = getNewStress(point);
    if (newStress >= 7) {
      context.scheduler().schedule(new Fall(pos));
    }

    if (point.stress() != newStress) {
      point.setStress(newStress);
      context.scheduler().schedule(new UpdateNeighborStress(pos));
    }
  }

  private static int getNewStress(PointedWorld point) {
    if (!point.isStressAware()) {
      return 0;
    }

    PointedWorld below = point.move(Direction.DOWN);
    int horizontalStress =
        Direction.Plane.HORIZONTAL
            .stream()
            .mapToInt(direction -> point.move(direction).stress() + 1)
            .min()
            .orElse(Integer.MAX_VALUE);

    return Math.min(horizontalStress, below.stress());
  }
}

File: src/main/kotlin/graveler/action/UpdateStress.kt
package graveler.action

import graveler.util.PointedWorld
import graveler.util.pointedAt
import graveler.util.stressMap
import net.minecraft.block.Blocks
import net.minecraft.util.Direction
import net.minecraft.util.math.BlockPos

data class UpdateStress(val pos: BlockPos) : Action {
  override val weight: Double = 0.0

  override fun apply(context: ActionContext) {
    val world = context.world
    val point = world.pointedAt(pos)

    if (!point.isStressAware) {
      point.clearStress()
      return
    }

    if (!world.isAreaLoaded(pos, 1)) {
      return
    }

    val newStress = point.getNewStress()
    if (newStress >= 7) {
      context.scheduler.schedule(Fall(pos))
    }

    if (point.stress != newStress) {
      point.stress = newStress
      context.scheduler.schedule(UpdateNeighborStress(pos))
    }
  }

  private fun PointedWorld.getNewStress(): Int {
    if (!isStressAware) return 0

    val below = move(Direction.DOWN)
    val horizontalStress = Direction.Plane.HORIZONTAL
      .map { move(it).stress + 1 }

    return (horizontalStress + below.stress).min()!!
  }

  private fun PointedWorld.clearStress() {
    chunk?.stressMap?.stresses?.remove(pos)
  }

  private var PointedWorld.stress: Int
    get() = when {
      isStressAware -> chunk?.stressMap?.get(pos) ?: 0
      isPermanentlyStable -> 0
      else -> MAX_STRESS
    }
    set(value) {
      when {
        isStressAware -> chunk?.stressMap?.set(pos, value)
      }
    }

  private val PointedWorld.isStressAware: Boolean
    get() = !blockState.isAir(world, pos) && !isPermanentlyStable

  private val PointedWorld.isPermanentlyStable: Boolean
    get() = block == Blocks.BEDROCK || material.isLiquid

  companion object {
    private const val MAX_STRESS = 50
  }
}
